pkg/kubelet/kubelet_test.go

```diff
@@ -1002,45 +1011,126 @@ func TestDeleteOutdatedMirrorPod(t *testing.T) {
 }
 
 func TestDeleteOrphanedMirrorPods(t *testing.T) {
-	testKubelet := newTestKubelet(t, false /* controllerAttachDetachEnabled */)
-	defer testKubelet.Cleanup()
-
-	kl := testKubelet.kubelet
-	manager := testKubelet.fakeMirrorClient
-	orphanPods := []*v1.Pod{
+	testCases := []struct {
+		// gracePeriod is the grace period for pod termination
+		// runningDuration is the duration for pod container to be in ContainerStateRunning. After that, container goes to ContainerStateExited
+		gracePeriod, runningDuration int64
+		qos                          bool
+	}{
 		{
-			ObjectMeta: metav1.ObjectMeta{
-				UID:       "12345678",
-				Name:      "pod1",
-				Namespace: "ns",
-				Annotations: map[string]string{
-					kubetypes.ConfigSourceAnnotationKey: "api",
-					kubetypes.ConfigMirrorAnnotationKey: "mirror",
-				},
-			},
+			gracePeriod:     0,
+			runningDuration: 0,
+			qos:             false,
 		},
 		{
-			ObjectMeta: metav1.ObjectMeta{
-				UID:       "12345679",
-				Name:      "pod2",
-				Namespace: "ns",
-				Annotations: map[string]string{
-					kubetypes.ConfigSourceAnnotationKey: "api",
-					kubetypes.ConfigMirrorAnnotationKey: "mirror",
-				},
-			},
+			gracePeriod:     2,
+			runningDuration: 1,
+			qos:             false,
+		},
+		{
+			gracePeriod:     2,
+			runningDuration: 1,
+			qos:             true,
 		},
 	}
+	for _, tc := range testCases {
+		pod := podWithUIDNameNsSpec("12345679", "pod1", "ns",
+			v1.PodSpec{
+				Containers: []v1.Container{
+					{Name: "1234", Image: "foo"},
+				},
+				TerminationGracePeriodSeconds: &tc.gracePeriod,
+			},
+		)
+		orphanPods := []*v1.Pod{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					UID:       "12345678",
+					Name:      "pod1",
+					Namespace: "ns",
+					Annotations: map[string]string{
+						kubetypes.ConfigSourceAnnotationKey: "api",
+						kubetypes.ConfigMirrorAnnotationKey: "mirror",
+					},
+				},
+			},
+		}
+		cgp := cm.CGroupPods{Pods: make(map[types.UID]cm.CgroupName), Lock: &sync.RWMutex{}}
+		cgp.Pods[orphanPods[0].UID] = []string{"kubepods", "burstable", "pod1234-abcd-5678-efgh"}
+		testKubelet := newTestKubeletWithCGroupPods(t, false /* controllerAttachDetachEnabled */, &cgp)
+		defer testKubelet.Cleanup()
 
-	kl.podManager.SetPods(orphanPods)
-	// Sync with an empty pod list to delete all mirror pods.
-	kl.HandlePodCleanups()
-	assert.Len(t, manager.GetPods(), 0, "Expected 0 mirror pods")
-	for _, pod := range orphanPods {
-		name := kubecontainer.GetPodFullName(pod)
-		creates, deletes := manager.GetCounts(name)
-		if creates != 0 || deletes != 1 {
-			t.Errorf("expected 0 creation and one deletion of %q, got %d, %d", name, creates, deletes)
+		kl := testKubelet.kubelet
+		manager := testKubelet.fakeMirrorClient
+
+		pods := []*v1.Pod{orphanPods[0], pod}
+		kl.cgroupsPerQOS = tc.qos
+		clock := kl.clock.(*clock.FakeClock)
+		kl.podManager.SetPods(pods)
+		kl.podManager.DeletePod(pod)
```

Comment:
its hard to follow what this test is trying to cover.  the cgroup pods stuff is not what is driving the core deletion behavior versus an implementation detail behind pod manager.  my concern is that when you call setPods() and then deletePod() you are actually changing the internal pod cached in the pod manager, but the core kubelet loop could just call `HandlePodUpdate` or `HandlePodReconcile` and you would lose all tracking state.  nothing is actually being introspected from the container runtime as manifested on disk that a container was pending deletion, instead everything is triggered from an internal cache getting updated for an object that is intended to be readonly by its consumers.

Reply:
The cgroup pods related code, cleanupOrphanedPodCgroups in particular, should also take into account whether the pod is in graceful termination before calling podContainerManagerImpl#Destroy.
Hence in the TestDeleteOrphanedMirrorPods, I have assertion that the underlying cgroup should (or should not) exist depending on the phase of the pod termination (determined by pod status).

I will add more comments to TestDeleteOrphanedMirrorPods outlining the purpose of each sub-test case.

w.r.t. container runtime, as far as I can tell, we may have to resort to certain stub for more assertions.
I am studying related tests to see if there is better way for writing new test.

Thanks